### ISSUE TITLE
Parser errors

### ISSUE THREAD

**Issue by schragge** [koalaman/shellcheck#449]
shellcheck 0.3.8 on Debian sid
```
In /usr/bin/doc line 121:
prepmenu () {
^-- SC1009: The mentioned parser error was in this function.
            ^-- SC1073: Couldn't parse this brace group.


In /usr/bin/doc line 142:
  eval r=({1..$menulen}); IFS=\|$IFS ptn="${r[*]}" IFS=${IFS#|}
         ^-- SC1056: Expected a '}'. If you have one, try a ; or \n in front of it.
         ^-- SC1036: '(' is invalid here. Did you forget to escape it?
         ^-- SC1072: Missing '}'. Fix any mentioned problems and try again.
```
The function in question:
```bash
prepmenu () {
# Parameters: none
# Global
#   uses: COLUMNS, pglen, idx, prompt{1,x,2}, p{1,x,2}len, ftype
#   sets: files, menu, extracase, ptn, cols, nwidth, page, pwidth, prompt
  local -i menulen linelen maxname ipp lastpage extralen hlen tlen
  local extra fspace hspace head tail sep sfx= swidth=

  local fraw file line
  # Since embedded newlines in file names are not impossible, make 'menu'
  # a copy of 'files' with newlines replaced by question marks.
  files=($idx) menu=("${files[@]//$'\n'/?}")
  while IFS= read -rd '' fraw; do
    file=${fraw:4} line=${file//$'\n'/?}${ftype[${fraw::1}]} linelen=${#line}
    ((linelen>maxname))&&maxname=$linelen
    files+=("$file") menu+=("$line")
  done < <(stat -L --printf=%f%n\\0 -- !($idx)|sort -z -k1.1,1.1)
  menulen=${#menu[@]}

  # Setting variables for printmenu
  local -a r
  eval r=({1..$menulen}); IFS=\|$IFS ptn="${r[*]}" IFS=${IFS#|}
  # Alternatively, $ptn may be initialized using 'seq'.
  #ptn=`seq -s'|' $menulen`
  nwidth=${#menulen}

  ((cols=COLUMNS/(nwidth+maxname+3) )) || cols=1
  ((cols+=menulen>pglen&&cols<2))
  ((ipp=cols*pglen))
  ((lastpage=menulen/ipp+!!(menulen%ipp) ))
  page=1

  if ((lastpage==1))
    then pwidth=. extra= extralen=0 extracase=
  else
    pwidth= extra=$promptx
    # Actually, 'extralen=13' would be enough in any real use scenario:
    #   on a standard 24-lines terminal 99 menu pages translate into at least
    #   4554 menu items (assuming 2-column display).
    let extralen=${#lastpage}+1+pxlen
    extracase='
      [tT]*) page=1;;
      [bB]*) page='$lastpage';;
      [nN]*) ((page+='$lastpage'>page));;
      [pP]*) ((page-=page>1));;
    '
  fi


  head=...${DIRSTACK[1]#$docpfx} hlen=${#head} tail=${PWD##*/} tlen=${#tail}
  ((hspace=(fspace=COLUMNS-p1len-extralen-p2len-nwidth-1)-tlen)) ||:
  sep=...
  if   (( hspace           >hlen     )); then sep=/          # head/tail
  elif (( (hlen=hspace-4)  >3        )); then sep=.../       # h.../tail
  elif (( (tlen=fspace-6)  >(hlen=0) )); then sfx=...        # ...t...
  elif (( (tlen+=${#extra})>0        )); then sfx=... extra= # ...t...
  # Given that $COLUMNS>60, the rest is extremely unlikely (>9999999 entries).
  elif (( (tlen=0)         >fspace-3 )); then sfx=' [...eXit...]> ' swidth=.
  fi
  printf -v prompt "%.${hlen}s$sep%.${tlen}s$sfx%${swidth}s" \
    "${head//$'\n'/?}" "${tail//$'\n'/?}" "$prompt1$extra$prompt2"
}
```
The whole script (doc.bash) may be downloaded here:
https://app.box.com/s/e0y3s4w4431z8u7e9e1w

**Comment by schragge** [comment]
> ```bash
> eval r=({1..$menulen}); IFS=\|$IFS ptn="${r[*]}" IFS=${IFS#|}
> ```

After giving it some more thought, now I think that instead of this monstrosity I should just have gone with *seq* like what the comment in the code suggests:
```bash
ptn=`seq -s'|'  $menulen`
```
Besides conveying my intent much more clearly, it also scales better. When doing this, I obviously was trying to avoid unnecessarily forking another process. While most packages come with only a few  files installed under /usr/share/doc, there're some like glibc-doc or texlive-doc with hundreds of directory entries. Brace expanding such a range, then saving it temporarily in an array just to interpolate it into a string with | as separator is probably less efficient than constructing the string with *seq*. Or, like GrayCat puts it in his excellent [BashPitfalls][BP33]:

> brace expansion pre-expands every argument which can be slower
> and unnecessarily consumes memory

[BP33]: http://mywiki.wooledge.org/BashPitfalls#for_i_in_.7B1...24n.7D "for i in {1..$n}"

Restricting it to only small values of $menulen like this
```bash
if ((menulen<20)); then
  local -a r
  eval r=({1..$menulen}); IFS=\|$IFS ptn="${r[*]}" IFS=${IFS#|}
  unset r
else
  ptn=`seq -s'|'  $menulen`
fi
```
seems like just complicating things for no apparent gain.

Nevertheless, evaluating brace expanded range to interpolate variable values seems like a legitimate bash idiom that should not give parser errors.

**Comment by schragge** [comment]
Ok, trying to isolate the problem
```bash
#!/bin/bash
set -e -x
len=7
declare -a r
eval r=({1..$len})
IFS=\|
echo "${r[*]}"
```
When the argument to *eval* is unquoted, shellcheck gives
```
In bin/yyy line 3:
len=7
^-- SC2034: len appears unused. Verify it or export it.


In bin/yyy line 4:
declare -a r
           ^-- SC2034: r appears unused. Verify it or export it.


In bin/yyy line 5:
eval r=({1..$len})
       ^-- SC1088: Parsing stopped here. Invalid use of parentheses?
       ^-- SC1036: '(' is invalid here. Did you forget to escape it?
```
However, quoting the argument like this:
```bash
eval "r=({1..$len})"
```
makes shellcheck happy. So, this obviously was not a parsing error after all, or rather the parser error had nothing to do with this construct. But that leaves the question, what was the parser error in the prepmenu() function declaration all about? Double-quoting the eval argument in the function, or replacing it altogether with the aforementioned *seq* expression still gives me
```
In src/deb/doc-1/doc.bash line 121:
prepmenu () {
            ^-- SC1009: The mentioned parser error was in this brace group.


In src/deb/doc-1/doc.bash line 150:
  ((lastpage=menulen/ipp+!!(menulen%ipp) ))
  ^-- SC1073: Couldn't parse this ((..)) command.
                          ^-- SC1072: Expected "--". Fix any mentioned problems and try again.
```
So, now we have this test case
```bash
#!/bin/bash
set -e -x
len=7 ipp=20
((lastpage=len/ipp+!!(len%ipp) ))
echo "lastpage=$lastpage"
```
Here, just adding another pair of parentheses makes the error go away:
```bash
((lastpage=len/ipp+!(!(len%ipp) ) ))
```
After this last change, I've got a bunch of shellcheck warnings, but they all are either yellow or green, mostly about using backticks instead of the POSIX $(...) command substitution, possible underquoting of variables, and initializing several variables to empty values in one line like this:
```bash
foo= bar= baz=
```
All those are quite understandable from the best coding practice POV that shellcheck tries to promote, and I mostly find them appropriate. All the red errors disappeared, and shellcheck was able to parse the code.

**Comment by thenoncoder** [comment]
￼

**Comment by koalaman** [comment]
These are some interesting cases!

`(( !!a ))` failing is a straight up parser bug: it has been fixed in 01c27dc. 

`eval` is weirder. By standard parsing rules, `eval foo=(bar)` would not be allowed. `bash` treats it specially and allows what would normally be valid array assignments. `ksh` additionally allows things like like `eval $var=(bar)`, which bash does not. 

Here's how different magic values affect parsing in `bash`:

    $ foo() { echo "$@"; }; eval() { echo "$@"; }; export() { echo "$@"; }
    $ foo a=(b)
    bash: syntax error near unexpected token `('
    $ eval a=(b)
    a=(b)
    $ export a=(b)
    a

7f547cc adds a warning saying to escape special characters to avoid relying on this fragile behavior.
